Improve AI MCV deployment

reaperrr opened the pull request:
While #11308 has merit of its own, it doesn't really solve #11305.

Instead, I took a closer look at the AI MCV deployment code. This PR fixes the two more problematic out of three issues I could find:
- The first is that the AI was spamming orders even to MCVs that were already on the move.
- The second is that the AI often chose locations near the enemy base, usually resulting in the MCV's destruction before it's even deployed.

The third is that `ChooseBuildLocation` works rather poorly in practice, sometimes choosing locations where some obstacle doesn't allow deployment, but fixing that will be harder and is therefore left for some other time.

The commits should be descriptive enough, so I won't repeat the details here.

reaperrr committed:
Fix AI giving orders to MCVs that already have orders

Checking for IsMoving is a flawed approach no matter how you look at it. The MCV might just have temporarily stopped due to an obstacle, or about to be deployed.

Checking for IsIdle instead ensures that the MCV really isn't already in the process of doing something.

reaperrr committed:
Make AI deploy MCVs in the vincinity of existing construction yards

While the general idea of AIs building distant secondary bases might be tempting, in reality the AI would way too often send the MCV close to some enemy base, wasting the cash and potentially increased build speed/additional build queue.

Deploying MCVs close to the existing base ensures that the AI will actually have some benefit from building an MCV.

abcdefg30 commented:
Won't this make the AI only deploy mcvs in their build radius?
Not sure if that is really beneficial (as the build radii overlap).

reaperrr commented:
> Won't this make the AI only deploy mcvs in their build radius?

Yes. In my opinion overlapping of build radi is a far lesser evil than never getting to deploy the MCV at all or lose the deployed yard too quickly, at least for the AI.

Deploying it within the build radius will still increase the maximum size of the base, increase build speed and make it easier to defend the new CYard compared to deploying it somewhere random and getting it destroyed/captured on its way there or shortly after.

If all player/AI slots are filled, the area around the own starting location is usually the sanest spot to deploy a second MCV anyway.


We can revisit this aspect once we come up with a smarter deploy location decision-making for the AI, but that is worth its own PR as it will likely require more work.

pchote commented:
> We can revisit this aspect once we come up with a smarter deploy location decision-making for the AI, but that is worth its own PR as it will likely require more work.

Sensible games allow map authors to add AI metadata for things like secondary base locations using waypoints or regions.  I've been saying for a long time that we should do this for OpenRA as part of our much needed AI overhaul.

reaperrr commented:
What we could do as a first step in that direction would be to make the AI check the spawnpoints on the map first, and if there are any with no enemies within a certain perimeter around that spawnpoint, send the MCV there, and if none are available, deploy it inside the own spawnpoint perimeter.

GraionDilach commented:
I might be alone here but I like the wandering mobile war factory mode (abusing the AI Mcvs list for MWFs), with considering only the "find-location-go-elsewhere-only-deploy-on-the-fifth-fitting-location" aspect as a bug, but such an additional waypoint system sounds good too.

yaaaaa commented:
I think that AI should build an MCV, even if it has a CONYARD.
And then deploy an MCV to another CONYARD next to the spawn resources area (oge mine or gem mine).
AI must do so prior to the moment when all the ore mines and/or gem mines areas will not be busy itself conyards.

pchote commented:
Needs a rebase now.

reaperrr commented:
Rebased.
I'm a bit low on head-bandwidth at the moment and according to https://github.com/OpenRA/OpenRA/blob/bleed/OpenRA.Mods.Common/AI/HackyAI.cs#L820 commit 2 is indeed a bugfix, so I'll stick to this for now and leave spawnpoint consideration or AI metadata for dedicated follow-up(s).

Nelax commented:
Much better logic. The AI tries to keep base integrity with deploying units. Maybe in the future the AI will make a calculated decision on whether to deploy the MCV near its base or not.
[img]http://imgur.com/xlW3dal.png[/img]

Phrohdoh commented:
I consider this a sane base expansion strategy. 👍 

```diff
diff --git a/mods/ra/rules/ai.yaml b/mods/ra/rules/ai.yaml
index bafed36..f322f25 100644
--- a/mods/ra/rules/ai.yaml
+++ b/mods/ra/rules/ai.yaml
@@ -43,24 +43,7 @@ Player:
 			fix: 0.1%
 			dome: 10%
 		UnitsToBuild:
-			e1: 50%
-			e2: 20%
-			e3: 10%
-			e4: 2%
-			dog: 2%
-			shok: 2%
-			harv: 10%
-			apc: 30%
-			jeep: 40%
-			arty: 15%
-			v2rl: 40%
-			ftrk: 50%
-			1tnk: 70%
-			2tnk: 25%
-			3tnk: 50%
-			4tnk: 10%
-			ttnk: 10%
-			stnk: 5%
+			mcv: 100%
 		UnitLimits:
 			dog: 4
 			harv: 8
```

![screen shot 2016-06-24 at 11 14 14 am](https://cloud.githubusercontent.com/assets/4861023/16343451/dd7fda90-39fc-11e6-9e37-a0fcb1093685.png)

reaperrr committed:
Allow modders to choose AI MCV deployment base restriction

For some mods, a random map location might be a better choice for the AI than deploying the MCV inside the base, so we allow modders to customize it.

reaperrr commented:
Updated. Made the restriction to base radius optional.

reaperrr commented:
I count a total of +3 so far and Nolt confimed on IRC that this fixes his problems with flying MCVs in his mod, so I'll merge it now.

reaperrr commented:
[changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/8963319f838e8f4efbe662ca82a062456ea1fe2f...7eb9f03c390e3c3843e8de9618c8282a46363545)
